### DISCUSSION TITLE
PWA é uma boa escolha?

### DISCUSSION THREAD

**Discussion by elenmagalhaes** [frontendbr/forum#1564] [Perguntas e Respostas]
Oi pessoal, tudo bem? 

Sou uma observadora assídua do fórum, mas nunca publiquei aqui, pois conheci o frontendbr recentemente (mal conheço, mas já considero pakas haha :stuck_out_tongue:). 
Antes de criar essa issue, dei uma olhada em bastante postagens, achei até algo 'similar' a minha dúvida [aqui](https://github.com/frontendbr/forum/issues/1434), mas não sei se é exatamente a solução para o que preciso, por isso gostaria da ajuda de vocês.

Meu cliente tem a necessidade hoje de criar um MVP, ou seja, inicialmente ele não investirá muito no projeto e nem em recursos. Em resumo, ele precisa de um site estático com algumas rotas básicas e sem autenticação (por enquanto), para que ele já possa ir trabalhando o marketing em cima disso.

Porém, num futuro próximo ele deseja transformar isso em um app (loja) e também incluir a parte de autenticação, crud, chatbot etc.

Então comecei a pensar na stack do projeto e lembrei do PWA. Ele me pareceu perfeito, tirando pelo fato que ele quer fazer o download do app na loja.

Nunca trabalhei com PWA. Na verdade, para dizer que nunca tive contato, eu participei de um projeto onde incluí o Google Analytics e só, rs. Então a partir disso, aqui estão minhas dúvidas:

1. Vocês acham uma boa ideia eu iniciar um projeto desse em PWA? Creio que se tiver muito ponto positivo em cima da utilização dele, eu possa convencê-lo a não ser um app de loja.

2. Fuçando por aí, encontrei [este repositório](https://github.com/felipefialho/kratos-boilerplate). Acham uma boa para eu dar um start?

3. Vocês poderiam me indicar bons conteúdos para estudo ou tutoriais de como aplicar? Pode ser conteúdo em inglês.
Obs.: Eu encontrei [esse artigo](https://willianjusten.com.br/como-fazer-seu-site-funcionar-offline-com-pwa/) do Willian Justen que achei sensacional, mas não queria ter apenas uma fonte.

4. Caso eu siga por essa solução, quais as maiores dificuldades técnicas que eu possa encontrar?

5. Por fim, se não for uma boa solução, o que eu poderia utilizar que faria mais sentido?

Desde já agradeço :hugs:

**Comment by eduardosoareszup** [comment]
Dei uma olhada rápida em seu Github e pelo que vi você possui conhecimento em Angular, correto?

Acho que no seu caso talvez seja interessante utilizar [Ionic](https://ionicframework.com/), pelos seguintes motivos:

- Como é um MVP o desenvolvimento vai acabar sendo mais rápido usando ele, visto que ele possui muitos recursos que você pode reaproveitar, então você poderá fazer algo mais geral nesse primeiro momento e depois ir fazendo as adequações que seu cliente for sugerindo;
- Com ele você pode gerar um App para Android/IOS, senão usar muitos recursos nativos você não terá muitos problemas e a performance poderá ser boa, fora que conseguirá colocar o App na loja.
- Segundo a lenda ele é capaz de gerar PWA também, nunca fiz isso e não conheço quem tenha gerado apenas a PWA, normalmente a galera usa ele para criar Aplicativos para colocar na loja mesmo, por isso eu falei lenda kk.
- Você possui conhecimento em Angular (correto?), então não terá muitas dificuldades em aprender a usar o Ionic, caso você nunca tenha usado.

Espero ter ajudado 🕺

**Comment by sbvorpagel** [comment]
Sim, é uma boa. Claro a PWA ainda possui algumas limtações (principalmente quando olhamos o Safari do Mac, no ios até que está andando).

Mas uma coisa boa é usar um projeto como o Capacitor (https://capacitor.ionicframework.com/) onde podemos gerar um apk ou ipa para publicarmos nas stores.

A stack não importa tanto assim, desde que a saída dela seja html, css e JS. Então pode ser Vue, Ionic, Angular, React ou mesmo JS puro (claro existem outras inúmeras).

**Comment by eduardosoareszup** [comment]
@sbvorpagel muito bacana cara! 😄

**Comment by felipefialho** [comment]
Primeiramente seja bem vinda @elenmagalhaes 😄 

Sobre sua dúvida, diria que sim, PWA é uma boa escolha para boa parte dos produtos atualmente. Apesar de ter algumas limitações com relação a um app comum (limitações que estão diminuindo cada vez mais), ainda sim é extremamente escalável e fácil de desenvolver.

Quanto a sua questão de publicar nas lojas, tenho uma ótima noticia: Isso já é possível já faz algum tempo na *Play Store* e foi uma das implementações que a Apple fez ano passado, então você consegue colocar seu PWA tanto na [Play Store quanto na Apple Store](https://hackernoon.com/progressive-web-apps-apple-app-store-google-play-store-it-was-nice-knowing-you-728a00350e67).

Esse [artigo do Will](https://willianjusten.com.br/como-fazer-seu-site-funcionar-offline-com-pwa/) é muito bom, e você pode pegar de base com certeza. 

E para finalizar, por coincidência o [Kratos Boilerplate](https://github.com/felipefialho/kratos-boilerplate) que você citou foi desenvolvido por mim,  😂. Posso dizer que sim, ele já gera um PWA automaticamente, mas foi pensado para projetos mais simples, de preferencia estáticos. 

No seu caso iria para uma solução mais robusta, olharia com muito carinho o [GatsbyJS](https://www.gatsbyjs.org/) e o [Next.js](https://nextjs.org/), tanto eu quanto o @willianjusten escrevemos artigos de como desenvolvemos nosso blog usando o Gatsby, sendo que o WIll lançou inclusive um curso sobre o assunto:

- [Willian Justen - Making of blog novo](https://willianjusten.com.br/making-of-blog-novo/)
- [Felipe Fialho - Como foi desenvolver meu novo blog usando o GatsbyJS](https://www.felipefialho.com/blog/como-foi-desenvolver-meu-novo-blog-usando-o-gatsbyjs/)

**Comment by felipefialho** [comment]
Vale olhar #1388

**Comment by lucas-moraes** [comment]
Oi @elenmagalhaes  tudo bem ?

Eu tenho um projeto chamado [https;//cloudcont.app](https://cloudcont.app), eu desenvolvi ele fullstack em linguagem HTML, CSS, Jquery, PHP e MYSQL. ele trabalha um pouco diferente dos PWA's normais, por que ao invés dele ficar responsivo, ele reconhece qual a fonte que está o acessando, e assim ele repassa o usuário para um site menor e mais leve, ou mantém o site desktop.
Eu na verdade sou contador mas amo Dev, e elaborei esse site por causa de um projeto start-up próprio.

Acho que uma das maiores dificuldades seria embutir todo um leque de conteúdo de desk para mobile, já que o projeto seria uma Loja virtual, e economizar tempo de carregamento junto com o trabalho off-line.

**Ionic seria uma boa por enquanto, já que o Flutter Web tá amadurecendo.**

**Comment by elenmagalhaes** [comment]
@eduardosoareszup,
Sim, eu desenvolvo em Angular. Inclusive, meu Github está uma bagunça haha. 
Já tinha ouvido falar do Ionic, mas nunca cheguei a pegar pra ler a função dele. Muito bom! Vou dar uma lida a fundo para entender melhor. Agradeço mt pela ajuda! :smile:

@sbvorpagel 
Cara, muito boa essa solução, muito boa mesmo, principalmente porque como já desenvolvo em Angular, parece muito mais simples pra mim. Vou pesquisar só se a implementação não é muito complexa e tentar dar uma brincada. Muito obrigada! :smile:

@felipefialho 
Puts, que boa notícia essa que é possível publicar na loja a partir do PWA! Porém tava dando uma lida e fiquei meio assustada com o processo de implementação para essa disponibilização. Me pareceu um pouco complexo, mas vou ler melhor, às vezes só foi o linguajar que assusta haha
Uma pena o seu boilerplate não ser muito adequado pro meu projeto, mas depois vou tentar fazer uma poc pra ver como funfa :stuck_out_tongue_winking_eye:
Obrigada pela indicação dos links. Vou dar uma lida a respeito.

@lucas-moraes 
Muito bom o site, eu achei super clean e elegante. Parabéns! 
Então, em relação a ser offline, não pesa tanto, não é um requisito, mas ser disponibilizado em loja sim. Apesar do PWA não ter essa possibilidade, como ele libera um ícone na tela inicial, acho que já é o suficiente pra ele.
Vou ler mais sobre Ionic. Me parece uma boa opção! Muito obrigada pelo tempo e ajuda. :smile:

**Comment by caioincau** [comment]
Meus 50 cents, acho PWA  uma boa, mas como dito, é bom compilar para colocar na loja, pensando em business a loja é algo ótimo para aquisição de clientes.

**Comment by onlurking** [comment]
Recentemente o Google lançou uma nova tecnologia chamada [TWA (Trusted Web Activities)](https://developers.google.com/web/updates/2019/02/using-twa) que basicamente você empacota teu site PWA dentro de um apk para distribuição, já tem alguns artigos de como fazer isso:
- [This TWA stuff rocks! Finally I got my PWA on Google Play Store 🎉](https://medium.com/@svenbudak/this-twa-stuff-rocks-finally-i-got-my-pwa-on-google-play-store-b92fe8dae31f)
- [Trusted Web Activity - PWA to Play Store Guide](https://fireship.io/lessons/pwa-to-play-store/)

**Comment by ninetails** [comment]
Apesar de recomendar também _TWA_, diria pra tentar ir pra Ionic que vai muito bem com Angular e talvez dê pra reutilizar código e expertise, levando em conta o que a OP conhece.
